lib/grpc/stream.ex

```diff
@@ -1,17 +1,433 @@
+    - `opts`: Optional keyword list for configuring the Flow or GenStage producer.
+
+  ## Behavior
+
+    - Automatically extracts gRPC metadata from the stream headers and injects it into the options as `:metadata`.
+
+  ## Returns
+
+    - A `GRPC.Stream` that emits the single gRPC message under demand.
+
+  ## Example
+
+      flow = GRPC.Stream.single_as_ctx(request, stream, max_demand: 10)
+
+  """
+  @spec single_as_ctx(any(), GRPC.Server.Stream.t(), Keyword.t()) :: t()
+  def single_as_ctx(input, %GRPC.Server.Stream{} = materializer, opts \\ [])
+      when is_struct(input) do
+    meta = get_headers(materializer) || %{}
+    single(input, Keyword.merge(opts, metadata: meta))
+  end
+
+  @doc """
+  Wraps an existing `Flow` into a `GRPC.Stream` struct.
+
+  > Note: `:join_with` is not supported in this mode and will raise an error if passed.
+
+  ## Parameters
+
+    - `flow`: A valid `Flow` pipeline.
+    - `opts`: Optional keyword options.
+
+  ## Returns
+
+    A `%GRPC.Stream{}` struct wrapping the flow.
+  """
+  @spec from_flow!(Flow.t(), Keyword.t()) :: t()
+  def from_flow!(%Flow{} = flow, opts \\ []) do
+    if Keyword.has_key?(opts, :join_with) do
+      raise ArgumentError, "join_with option is not supported for Flow input"
+    end
+
+    %__MODULE__{flow: flow, options: opts}
+  end
+
+  @doc """
+  Extracts the internal `Flow` pipeline from a `GRPC.Stream`.
+
+  Raises an `ArgumentError` if the `flow` has not been initialized.
+
+  ## Returns
+
+    The internal `Flow` struct.
+  """
+  @spec to_flow!(t()) :: Flow.t()
+  def to_flow!(%__MODULE__{flow: nil}) do
+    raise ArgumentError, "GRPC.Stream has no flow, initialize it with from/2"
+  end
+
+  def to_flow!(%__MODULE__{flow: flow}), do: flow
+
+  @doc """
+  Executes the flow and emits responses into the provided gRPC server stream.
+
+  ## Parameters
+
+    - `flow`: A `GRPC.Stream` struct containing the flow to be executed.
+    - `stream`: A `GRPC.Server.Stream` to which responses are sent.
+    - `opts`: Optional keyword list. Currently supported:
+    - `:dry_run` — If `true`, responses are not sent (used for testing or inspection).
+
+  ## Returns
+
+    - `:ok` if the stream was processed successfully.
+    - For unary mode, returns the first item in the stream instead of sending responses.
+
+  ## Example
+
+      GRPC.Stream.run_with(request, mat)
+  """
+  @spec run_with(t(), Stream.t(), Keyword.t()) :: :ok | any()
+  def run_with(
+        %__MODULE__{flow: flow, options: flow_opts} = _stream,
+        %Materializer{} = from,
+        opts \\ []
+      ) do
+    dry_run? = Keyword.get(opts, :dry_run, false)
+    unary? = Keyword.get(flow_opts, :unary, false)
+
+    if unary? do
+      flow
+      |> Enum.to_list()
+      |> List.flatten()
+      |> List.first()
+    else
+      flow
+      |> Flow.map(fn msg ->
+        unless dry_run?, do: send_response(from, msg)
+        flow
+      end)
+      |> Flow.run()
```

Comment:
I think dry_run would be cleaner if extracted to the outer if-else:

```elixir
cond do
  dry_run? ->
    ...
  unary? ->
    ...
   true -> 
     ...
end
```

This also makes dry-run work for unary flows as well, especially if we make the result value of an uniform type.

Reply:
I removed the polymorphic return